### PR TITLE
[DataLoader2] Saving and restoring initial seed generator

### PR DESCRIPTION
**Pull request by NivekT**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #915
* #1056
* __->__ #998

Changes to `DataLoader2`:
- Modifying `state_dict` to store `randomness_state`, which includes:
    - `_seed: int`
    - `_reset_seed: bool` - flag indicating whether `_seed` needs to be set
    - `_seed_generator` - the latest version at the time when `state_dict` is called
    - `_initial_seed_generator` - the versopm that is saved at the beginning of very epoch
- Modifying `from_state` and `load_state_dict` to restore `randomness_state`
- Adding a method `_restore_checkpoint_beginning_of_epoch`
    -  This sets `self._seed_generator = self._initial_seed_generator`, allowing users to re-create an epoch from the beginning.

---
### Considerations

Storing the randomness states provide more flexibility for users to restore as they see fit. The decision to do that should not be controversial.

I decided to make add a new method for checkpointing at the beginning of the epoch, ensure that users are not confused about what randomness is restored by default.

The basic idea is that we want to allow users to restore `dl2._seed_generator` to the previously saved version. From that point on, they can create a new `__iter__` and continue from the beginning of the epoch.
- Note that since `_seed` and `_reset_seed` are also saved, if the users were planning to use a different seed or if there was a need to re-seed, those remain valid after restoring the checkpoint. 
- Finally, if users change their mind at any point (after restoring) and want to manual set `seed`. That `seed` will override any other behavior and the `seed` will be used.

Differential Revision: [D44390519](https://our.internmc.facebook.com/intern/diff/D44390519)

### CODE REVIEW

File: torchdata/dataloader2/dataloader2.py [hunk 1]
```diff
@@ -206,9 +212,14 @@ def __iter__(self) -> DataLoader2Iterator[T_co]:
                 if self._reset_seed:
                     self._seed_generator.seed(self._seed)
                     self._reset_seed = False
```
Comment (NivekT) [inline torchdata/dataloader2/dataloader2.py:214]: In this branch of logic, we have `self._seed` and we use it to set. Should we set `self._seed = None` afterwards?
Currently, it is relying on the fact that `self._reset_seed == False` until `self.seed()` method is called again.

**Review by NivekT — COMMENTED** [review]
I decided to make modification to the existing APIs. Alternatively, we can create a new method.

The basic idea is that we want to allow users to restore `dl2._seed_generator` to the previously saved version, at the same time, we need to skip over the logic that re-do seeding in `initialize_iteration` (hence the new variable `_skip_iteration_seeding` is needed.

I see 2 main scenarios:
1. Users want to restore DataPipe and ReadingService but not the initial state of RNG
    - I think lots of current users (including some internals) are in this category.
    - This should work by default, because `restore_initial_seed_generator=False`, things will work the same as before
2. Users actively want to restore DP, RS, and initial state of RNG
    - Users will need to set an extra variable to `True` and we will make sure `_skip_iteration_seeding=True` so no re-seeding will happen in the first subsequent call of `initialize_iteration`

Finally, if users change their mind at any point (after restoring) and want to manual set `seed`. That `seed` will override any other behavior and the `seed` will be used.

**Review by ejguan — COMMENTED** [review]
I have two main comments:
- `SeedGenerator` is pickable. No need to add those supporting functions.
- IIUC, initial state should be the state at the beginning of each epoch but not the beginning of all epochs

File: torchdata/dataloader2/dataloader2.py [hunk 2]
```diff
@@ -280,17 +292,20 @@ def state_dict(self) -> Dict[str, Any]:
```
Comment (NivekT) [inline torchdata/dataloader2/dataloader2.py:292]: I think at it's current form, we want to encourage people to call `state_dict` at the beginning of the epoch? Or, we need to make sure nothing within each DataPipe's buffer is serialized (which I think is already true for now)? Thoughts?

The changes may come in subsequent PR

**Comment by NivekT** [comment]
> * `SeedGenerator` is pickable. No need to add those supporting functions.

Yea, that makes sense. I can take those out.

> * IIUC, initial state should be the state at the beginning of each epoch but not the beginning of all epochs

I'm expecting `__iter__` to be called at the beginning of each epoch, such that this line will be called every time.
`self._initial_seed_generator = clone(self._seed_generator)`

Does that look right to you?

**Comment by ejguan** [comment]
I am not sure if it's needed to differentiate initial seed and current state of seed generator. After SG seeds the graph, the state of SG is not going to evolve during iteration. All random states should be self-preserved by random DataPipes.
-> It means I don't know if we need to add `restore_initial_seed_generator` to the argument.

During `__iter__`, like what we offers for `reading_service`, whenever `seed_generator_state` is provided, we can just restore it. Otherwise, we should try to seed the `seed_generator` either by `self._seed` or `None`.

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43644876).

**Comment by NivekT** [comment]
I looked into whether we can refactor `._reset_seed:` from `DataLoader2` into `SeedGenerator`, I concluded that is not possible, mainly because the current DL2 behavior wants to use the seed once and don't re-seed until users call the `seed` method again. Whereas `SeedGenerator` would re-seed every time its `seed` method is called.

File: test/dataloader2/test_mprs.py [hunk 3]
```diff
@@ -197,6 +197,65 @@ def test_reading_service_limit(self, dp, n_workers, worker_prefetch_cnt, main_pr
...
+        self.assertEqual(list(it1), list(restored_dl))
+
+        dl.shutdown()
+        restored_dl.shutdown()
+
+        # Functional Test: Saving state after iterator is created and began iterating
+        dl = DataLoader2(datapipe=dp, reading_service=rs)
+        dl.seed(1)
+        it1 = iter(dl)
+        next(it1)  # Starts iterating
```
Comment (ejguan) [inline test/dataloader2/test_mprs.py:232]: Should we keep the result from `next(it1)`?

File: test/dataloader2/test_mprs.py [hunk 4]
```diff
@@ -197,6 +197,65 @@ def test_reading_service_limit(self, dp, n_workers, worker_prefetch_cnt, main_pr
...
+        restored_dl2._restore_checkpoint_beginning_of_epoch()
+
+        res1 = list(it1)
+        res2 = list(restored_dl)
+        res3 = list(restored_dl2)
+
+        print(res1)
+        print(res2)
+        print(res3)
+        # self.assertEqual(list(it1), list(restored_dl)[1:])  # Note skipping over 1st element from actual result
```
Comment (ejguan) [inline test/dataloader2/test_mprs.py:253]: nit: uncomment here  or do something like
```py
elem1 = next(it1)
...
self.assertEqual([elem1] + list(it1), list(restored_dl))
```

**Review by ejguan — APPROVED** [review]
Overall LGTM with a few comments below

File: test/dataloader2/test_mprs.py [hunk 5]
```diff
@@ -197,6 +197,65 @@ def test_reading_service_limit(self, dp, n_workers, worker_prefetch_cnt, main_pr
...
+        restored_dl2 = DataLoader2.from_state(initial_state, rs)  # type: ignore[arg-type]
+        restored_dl2._restore_checkpoint_beginning_of_epoch()
+
+        res1 = list(it1)
+        res2 = list(restored_dl)
+        res3 = list(restored_dl2)
+
+        print(res1)
+        print(res2)
+        print(res3)
```
[the remark this answers is not in the document]
Comment (NivekT) [reply test/dataloader2/test_mprs.py:252]: @ejguan Sorry this comment was pending and didn't get published

[the remark this answers is not in the document]
Comment (ejguan) [reply test/dataloader2/test_mprs.py:252]: Is the length of `res1` same as `res2`? If so, `res2` has one missing sample, right?

[the remark this answers is not in the document]
Comment (NivekT) [reply test/dataloader2/test_mprs.py:252]: `res2` and `res3` are longer by one (as expected). I am more concerned the ordering are exactly the same. I was expecting `res2`'s order to be different

[the remark this answers is not in the document]
Comment (ejguan) [reply test/dataloader2/test_mprs.py:252]: When `_restore_checkpoint_beginning_of_epoch` is not called, can you pls check the value of `self._reset_iter`, `self._seed` and `self._reset_seed` from [here](https://github.com/pytorch/data/blob/e78ab6c9ec94f05f0a350ced7fe571f6863c20ec/torchdata/dataloader2/dataloader2.py#L193-L199)?

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44390519).

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44390519).

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44390519).
